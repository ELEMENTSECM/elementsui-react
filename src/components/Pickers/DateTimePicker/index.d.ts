import * as React from 'react';

export type DateTimePickerValue = Date | string;

export type DateTimePickerDefaultValue = Date | string;

export type DateTimePickerViewDate = Date | string;

export type DateTimePickerDateFormat = 'L' | 'LL' | 'LLL' | 'LLLL' | 'LS' | 'LTS';

export type DateTimePickerTimeFormat = 'LT' | 'LTS';

export type DateTimePickerViewMode = 'years' | 'months' | 'days' | 'time';

export type DateTimePickerDateFrom = Date | string;

export type DateTimePickerDateTo = Date | string;

export interface DateTimePickerProps {
	/**
	 * HTML id tag of the root element
	 */
	id?: string;
	/**
	 * Input's 'name' attribute
	 */
	name?: string;
	/**
	 * Represents the selected date by the component, in order to use it as a controlled component. This prop is parsed by Moment.js, so it is possible to use a date string or a moment object.
	 */
	value?: DateTimePickerValue;
	/**
	 * Represents the selected date for the component to use it as a uncontrolled component. This prop is parsed by Moment.js, so it is possible to use a date string or a moment object.
	 */
	defaultValue?: DateTimePickerDefaultValue;
	/**
	 * Represents the month which is viewed on opening the calendar when there is no selected date. This prop is parsed by Moment.js, so it is possible to use a date string or a moment object.
	 */
	viewDate?: DateTimePickerViewDate;
	/**
	 * Defines the format for the date.
	 */
	dateFormat?: DateTimePickerDateFormat;
	/**
	 * Defines the format for the time.
	 */
	timeFormat?: DateTimePickerTimeFormat;
	/**
	 * Manually set the locale for the react-datetime instance. Moment.js locale needs to be loaded to be used, see i18n docs.
	 */
	locale: string;
	/**
	 * Callback trigger when the date changes. The callback receives the selected moment object as only parameter, if the date in the input is valid. If the date in the input is not valid, the callback receives the value of the input (a string).
	 */
	onChange?: (...args: any[]) => any;
	/**
	 * The default view to display when the picker is shown
	 */
	viewMode?: DateTimePickerViewMode;
	/**
	 * Extra class name for the outermost markup element.
	 */
	className?: string;
	/**
	 * Allowed dates from defined value
	 */
	dateFrom?: DateTimePickerDateFrom;
	/**
	 * Allowed dates to defined value
	 */
	dateTo?: DateTimePickerDateTo;
	/**
	 * Input is disabled
	 */
	disabled?: boolean;
	/**
	 * Input is required
	 */
	required?: boolean;
}

declare const DateTimePicker: React.SFC<DateTimePickerProps>;

export default DateTimePicker;
