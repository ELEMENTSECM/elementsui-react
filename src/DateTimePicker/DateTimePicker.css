.react-datepicker__triangle {
	transform: translate3d(78px, 0px, 0px) !important;
}
.react-datepicker__time-container--with-today-button {
	right: -87px !important;
}
