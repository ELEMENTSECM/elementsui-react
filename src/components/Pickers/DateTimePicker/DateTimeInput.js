import * as React from "react";
import { InputGroup, FormControl } from "react-bootstrap";

export default class DateTimeInput extends React.PureComponent {
	render() {
		const { onClick, disabled, showTimeSelectOnly, ...rest } = this.props;
		return (
			<InputGroup onClick={onClick}>
				<FormControl type="text" disabled={disabled} {...rest} />
				<InputGroup.Addon onClick={this.onIconClick} style={{ cursor: disabled ? "not-allowed" : "pointer" }}>
					<i className={`glyphicons glyphicons-${showTimeSelectOnly ? "clock" : "calendar"}`} />
				</InputGroup.Addon>
			</InputGroup>
		);
	}
}
