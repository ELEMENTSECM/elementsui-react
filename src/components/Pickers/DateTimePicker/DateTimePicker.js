import * as React from "react";
import PropTypes from "prop-types";
import Datetime from "react-datetime";
import { injectIntl, IntlProvider } from "react-intl";
import nls from "./DateTimePicker.nls.json";
import { InputGroup } from "react-bootstrap";

class DateTimePicker extends React.Component {
	constructor(props) {
		super(props);
		this.input = React.createRef();
	}

	calculalateValidDates = current => {
		const dateFromValid =
			!this.props.dateFrom || current.isAfter(Datetime.moment(this.props.dateFrom).subtract(1, "day"));
		const dateToValid = !this.props.dateTo || current.isBefore(Datetime.moment(this.props.dateTo).add(1, "day"));

		return dateFromValid && dateToValid;
	};

	onBlur = () => {
		this.props.onBlur && this.props.onBlur(this.input.current.state.selectedDate);
	};

	onIconClick = () => {
		this.input.current.openCalendar();
	};

	render() {
		const { name, disabled, required, dateFormat, timeFormat, viewMode, onBlur, ...rest } = this.props;
		const DTPicker = injectIntl(({ intl }) => {
			let dateFormatId = false,
				timeFormatId = false;

			if (viewMode !== "time") {
				if (dateFormat === undefined) {
					dateFormatId = intl.formatMessage({ id: "date" });
				} else if (typeof dateFormat === "string") {
					dateFormatId = intl.formatMessage({ id: `longDateFormat_${dateFormat}` });
				}
			}

			if (timeFormat === undefined) {
				timeFormatId = intl.formatMessage({ id: "time" });
			} else if (typeof timeFormat === "string") {
				timeFormatId = intl.formatMessage({ id: `longDateFormat_${timeFormat}` });
			}

			return (
				<InputGroup>
					<Datetime
						{...rest}
						ref={this.input}
						isValidDate={this.calculalateValidDates}
						inputProps={{ name, disabled, required }}
						dateFormat={dateFormatId}
						timeFormat={timeFormatId}
						onBlur={this.onBlur}
					/>
					<InputGroup.Addon onClick={() => this.onIconClick()} style={{ cursor: "pointer" }}>
						<i className={`glyphicons glyphicons-${viewMode === "time" ? "clock" : "calendar"}`} />
					</InputGroup.Addon>
				</InputGroup>
			);
		});

		return (
			<IntlProvider locale={this.props.locale} messages={nls[this.props.locale]}>
				<div id={this.props.id} className={this.props.className}>
					<DTPicker />
				</div>
			</IntlProvider>
		);
	}
}

DateTimePicker.propTypes = {
	/** HTML id tag of the root element */
	id: PropTypes.string,
	/** Input's 'name' attribute */
	name: PropTypes.string,
	/** Represents the selected date by the component, in order to use it as a controlled component. This prop is parsed by Moment.js, so it is possible to use a date string or a moment object. */
	value: PropTypes.oneOfType([ PropTypes.instanceOf(Date), PropTypes.string ]),
	/** Represents the selected date for the component to use it as a uncontrolled component. This prop is parsed by Moment.js, so it is possible to use a date string or a moment object. */
	defaultValue: PropTypes.oneOfType([ PropTypes.instanceOf(Date), PropTypes.string ]),
	/** Represents the month which is viewed on opening the calendar when there is no selected date. This prop is parsed by Moment.js, so it is possible to use a date string or a moment object. */
	viewDate: PropTypes.oneOfType([ PropTypes.instanceOf(Date), PropTypes.string ]),
	/** Defines the format for the date. */
	dateFormat: PropTypes.oneOfType([ PropTypes.oneOf([ "L", "LL", "LLL", "LLLL", "LT", "LTS" ]), PropTypes.bool ]),
	/** Defines the format for the time. */
	timeFormat: PropTypes.oneOfType([ PropTypes.oneOf([ "LT", "LTS" ]), PropTypes.bool ]),
	/** Manually set the locale for the react-datetime instance. Moment.js locale needs to be loaded to be used, see i18n docs. */
	locale: PropTypes.string.isRequired,
	/** Callback trigger when the date changes. The callback receives the selected moment object as only parameter, if the date in the input is valid. If the date in the input is not valid, the callback receives the value of the input (a string). */
	onChange: PropTypes.func,
	/** Callback trigger for when the user clicks outside of the input, simulating a regular onBlur. The callback receives the selected moment object as only parameter, if the date in the input is valid. If the date in the input is not valid, the callback returned. */
	onBlur: PropTypes.func,
	/** The default view to display when the picker is shown  */
	viewMode: PropTypes.oneOf([ "years", "months", "days", "time" ]),
	/** Extra class name for the outermost markup element. */
	className: PropTypes.string,
	/** Allowed dates from defined value */
	dateFrom: PropTypes.oneOfType([ PropTypes.instanceOf(Date), PropTypes.string ]),
	/** Allowed dates to defined value */
	dateTo: PropTypes.oneOfType([ PropTypes.instanceOf(Date), PropTypes.string ]),
	/** Input is disabled */
	disabled: PropTypes.bool,
	/** Input is required */
	required: PropTypes.bool,
	/** When true, once the day has been selected, the datepicker will be automatically closed. */
	closeOnSelect: PropTypes.bool
};

export default DateTimePicker;
