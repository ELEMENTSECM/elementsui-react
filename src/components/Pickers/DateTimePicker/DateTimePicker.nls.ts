export const nls = {
	en: {
		moment_g: "DD/MM/YYYY HH:mm",
		invalidDate: "Invalid date",
		longDateFormat_L: "DD/MM/YYYY",
		longDateFormat_LL: "MMMM Do YYYY",
		longDateFormat_LLL: "MMMM Do YYYY LT",
		longDateFormat_LLLL: "dddd, MMMM Do YYYY LT",
		longDateFormat_LT: "h:mm A",
		longDateFormat_LTS: "h:mm:ss A",
		date: "DD/MM/YYYY",
		time: "HH:mm:ss",
		datetime: "DD/MM/YYYY HH:mm",
		fullHour: "DD/MM/YYYY HH:00P"
	},
	nb: {
		moment_g: "DD.MM.YYYY HH:mm",
		invalidDate: "Ugyldig dato",
		longDateFormat_L: "DD.MM.YYYY",
		longDateFormat_LL: "D. MMMM YYYY",
		longDateFormat_LLL: "D. MMMM YYYY [kl.] LT",
		longDateFormat_LLLL: "dddd D. MMMM YYYY [kl.] LT",
		longDateFormat_LT: "H.mm",
		longDateFormat_LTS: "h:mm:ss",
		date: "DD.MM.YYYY",
		time: "HH:mm:ss",
		datetime: "DD.MM.YYYY HH:mm",
		fullHour: "DD.MM.YYYY HH:00P"
	},
	nn: {
		moment_g: "DD.MM.YYYY HH:mm",
		invalidDate: "Ugyldig dato",
		longDateFormat_L: "DD.MM.YYYY",
		longDateFormat_LL: "D. MMMM YYYY",
		longDateFormat_LLL: "D. MMMM YYYY [kl.] LT",
		longDateFormat_LLLL: "dddd D. MMMM YYYY [kl.] LT",
		longDateFormat_LT: "H.mm",
		longDateFormat_LTS: "h:mm:ss",
		date: "DD.MM.YYYY",
		time: "HH:mm:ss",
		datetime: "DD.MM.YYYY HH:mm",
		fullHour: "DD.MM.YYYY HH:00P"
	},
	sv: {
		moment_g: "YYYY-MM-DD HH:mm",
		invalidDate: "Ugyldig datum",
		longDateFormat_L: "YYYY-MM-DD",
		longDateFormat_LL: "D MMMM YYYY",
		longDateFormat_LLL: "D MMMM YYYY LT",
		longDateFormat_LLLL: "dddd D MMMM YYYY LT",
		longDateFormat_LT: "HH:mm",
		longDateFormat_LTS: "HH:mm:ss",
		date: "YYYY-MM-DD",
		time: "HH:mm:ss",
		datetime: "YYYY-MM-DD HH:mm",
		fullHour: "YYYY-MM-DD HH:00P"
	}
};
