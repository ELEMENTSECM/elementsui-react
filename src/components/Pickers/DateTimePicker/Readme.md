DateTimePicker date only:

```js 
<DateTimePicker
			locale="en"
			dateFormat="L"
			timeFormat={false}
			defaultValue={new Date()}
			onChange={(value) => alert(value)}
		/>
```
DateTimePicker time only:
```js
<DateTimePicker
				locale="en"
				viewMode="time"
				onChange={(value) => alert(value)}
				timeFormat="LT"
			/>
```