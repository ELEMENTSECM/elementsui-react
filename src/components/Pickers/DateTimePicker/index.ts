export { default, IDateTimePickerProps } from "./DateTimePicker";
