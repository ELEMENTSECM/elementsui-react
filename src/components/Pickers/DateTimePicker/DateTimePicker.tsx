import * as React from "react";
import Datetime from "react-datetime-contrib";
import { injectIntl, IntlProvider } from "react-intl";
import { nls } from "./DateTimePicker.nls";

export type DateTimePickerValue = Date | string;

export type DateTimePickerDefaultValue = Date | string;

export type DateTimePickerViewDate = Date | string;

export type DateTimePickerDateFormat = "L" | "LL" | "LLL" | "LLLL" | "LT" | "LTS" | boolean;

export type DateTimePickerTimeFormat = "LT" | "LTS" | boolean;

export type DateTimePickerViewMode = "years" | "months" | "days" | "time";

export type DateTimePickerDateFrom = Date | string;

export type DateTimePickerDateTo = Date | string;

export interface IDateTimePickerProps {
	/**
	 * HTML id tag of the root element
	 */
	id?: string;
	/**
	 * Input's 'name' attribute
	 */
	name?: string;
	/**
	 * Represents the selected date by the component, in order to use it as a controlled component. This prop is parsed by Moment.js, so it is possible to use a date string or a moment object.
	 */
	value?: DateTimePickerValue;
	/**
	 * Represents the selected date for the component to use it as a uncontrolled component. This prop is parsed by Moment.js, so it is possible to use a date string or a moment object.
	 */
	defaultValue?: DateTimePickerDefaultValue;
	/**
	 * Represents the month which is viewed on opening the calendar when there is no selected date. This prop is parsed by Moment.js, so it is possible to use a date string or a moment object.
	 */
	viewDate?: DateTimePickerViewDate;
	/**
	 * Defines the format for the date.
	 */
	dateFormat?: DateTimePickerDateFormat;
	/**
	 * Defines the format for the time.
	 */
	timeFormat?: DateTimePickerTimeFormat;
	/**
	 * Manually set the locale for the react-datetime instance. Moment.js locale needs to be loaded to be used, see i18n docs.
	 */
	locale: string;
	/**
	 * Callback trigger when the date changes. The callback receives the selected moment object as only parameter, if the date in the input is valid. If the date in the input is not valid, the callback receives the value of the input (a string).
	 */
	onChange?: (...args: any[]) => any;
	/**
	 * Callback trigger for when the user clicks outside of the input, simulating a regular onBlur. The callback receives the selected moment object as only parameter, if the date in the input is valid. If the date in the input is not valid, the callback returned.
	 */
	onBlur?: (...args: any[]) => any;
	/**
	 * The default view to display when the picker is shown
	 */
	viewMode?: DateTimePickerViewMode;
	/**
	 * Extra class name for the outermost markup element.
	 */
	className?: string;
	/**
	 * Allowed dates from defined value
	 */
	dateFrom?: DateTimePickerDateFrom;
	/**
	 * Allowed dates to defined value
	 */
	dateTo?: DateTimePickerDateTo;
	/**
	 * Input is disabled
	 */
	disabled?: boolean;
	/**
	 * Input is required
	 */
	required?: boolean;
}

class DateTimePicker extends React.Component<IDateTimePickerProps> {
	private input: React.RefObject<Datetime>;

	constructor(props: IDateTimePickerProps) {
		super(props);
		this.input = React.createRef();
	}

	calculalateValidDates = (current) => {
		const dateFromValid =
			!this.props.dateFrom || current.isAfter(Datetime["moment"](this.props.dateFrom).subtract(1, "day"));
		const dateToValid = !this.props.dateTo || current.isBefore(Datetime["moment"](this.props.dateTo).add(1, "day"));

		return dateFromValid && dateToValid;
	};

	onBlur = () => {
		// tslint:disable-next-line:no-unused-expression
		this.props.onBlur && this.props.onBlur(this.input.current.state.selectedDate);
	};

	render() {
		const { name, disabled, required, dateFormat, timeFormat, viewMode, onBlur, onChange, ...rest } = this.props;
		const DTPicker = injectIntl(({ intl }) => {
			let dateFormatId: string | boolean = false;
			let timeFormatId: string | boolean = false;

			if (viewMode !== "time") {
				if (dateFormat === undefined) {
					dateFormatId = intl.formatMessage({ id: "date" });
				} else if (typeof dateFormat === "string") {
					dateFormatId = intl.formatMessage({ id: `longDateFormat_${dateFormat}` });
				}
			}

			if (timeFormat === undefined) {
				timeFormatId = intl.formatMessage({ id: "time" });
			} else if (typeof timeFormat === "string") {
				timeFormatId = intl.formatMessage({ id: `longDateFormat_${timeFormat}` });
			}

			return (
				<Datetime
					{...rest}
					ref={this.input}
					isValidDate={this.calculalateValidDates}
					inputProps={{ name, disabled, required }}
					dateFormat={dateFormatId}
					timeFormat={timeFormatId}
					onBlur={this.onBlur}
				/>
			);
		});

		return (
			<IntlProvider locale={this.props.locale} messages={nls[this.props.locale]}>
				<div id={this.props.id} className={this.props.className}>
					<DTPicker />
				</div>
			</IntlProvider>
		);
	}
}

export default DateTimePicker;
