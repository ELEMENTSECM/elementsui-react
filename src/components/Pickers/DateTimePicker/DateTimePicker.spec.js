import * as React from 'react';
import { shallow } from 'enzyme';
import DateTimePicker from './index';

describe('DateTimePicker', () => {
	test('displays correctly', () => {
		function addDays(date, days) {
			var result = new Date(date);
			result.setDate(result.getDate() + days);
			return result;
		}

		const dateFrom = new Date('01/01/2018');
		const dateTo = addDays(dateFrom, 10);
		const wrapper = shallow(
			<DateTimePicker
				id="testDatePicker"
				locale="en"
				dateFrom={dateFrom}
				dateTo={dateTo}
				dateFormat={'LLL'}
				timeFormat={'LT'}
			/>
		);
		expect(wrapper).toMatchSnapshot();
	});
});
