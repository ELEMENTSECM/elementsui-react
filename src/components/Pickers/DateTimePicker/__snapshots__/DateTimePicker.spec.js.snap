// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateTimePicker displays correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <DateTimePicker
    dateFormat="LLL"
    dateFrom={2017-12-31T22:00:00.000Z}
    dateTo={2018-01-10T22:00:00.000Z}
    id="testDatePicker"
    locale="en"
    timeFormat="LT"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <div
        className={undefined}
        id="testDatePicker"
      >
        <InjectIntl(Component) />
      </div>,
      "locale": "en",
      "messages": Object {
        "date": "DD/MM/YYYY",
        "datetime": "DD/MM/YYYY HH:mm",
        "fullHour": "DD/MM/YYYY HH:00P",
        "invalidDate": "Invalid date",
        "longDateFormat_L": "DD/MM/YYYY",
        "longDateFormat_LL": "MMMM Do YYYY",
        "longDateFormat_LLL": "MMMM Do YYYY LT",
        "longDateFormat_LLLL": "dddd, MMMM Do YYYY LT",
        "longDateFormat_LT": "h:mm A",
        "longDateFormat_LTS": "h:mm:ss A",
        "moment_g": "DD/MM/YYYY HH:mm",
        "time": "HH:mm:ss",
      },
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <InjectIntl(Component) />,
        "className": undefined,
        "id": "testDatePicker",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": "div",
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <div
          className={undefined}
          id="testDatePicker"
        >
          <InjectIntl(Component) />
        </div>,
        "locale": "en",
        "messages": Object {
          "date": "DD/MM/YYYY",
          "datetime": "DD/MM/YYYY HH:mm",
          "fullHour": "DD/MM/YYYY HH:00P",
          "invalidDate": "Invalid date",
          "longDateFormat_L": "DD/MM/YYYY",
          "longDateFormat_LL": "MMMM Do YYYY",
          "longDateFormat_LLL": "MMMM Do YYYY LT",
          "longDateFormat_LLLL": "dddd, MMMM Do YYYY LT",
          "longDateFormat_LT": "h:mm A",
          "longDateFormat_LTS": "h:mm:ss A",
          "moment_g": "DD/MM/YYYY HH:mm",
          "time": "HH:mm:ss",
        },
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <InjectIntl(Component) />,
          "className": undefined,
          "id": "testDatePicker",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": "div",
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
